### Brief Introduction to `pandas`

- `pandas` is the defacto package for working with tabular data.
  - Think of it as `Excel` on steroids (without the graphical interface and the menus).
- Supports a plethora of file formats (`Excel`, `CSV`, `TSV`, `JSON`, `hdf5`, ....)


### Brief Introduction to Pandas - Cont'd

![](https://www.dropbox.com/s/vi451zvf9sw8jfr/pandas_architecture.png?dl=1)


### Pandas `Series` and `DataFrames`

* Available with the Conda Jupyter installation.

* `pandas` relies principally on two types of data structures:


1. `Series`: those are list-like objects that store data in a given order.
    * It helps to think of `Series` as a column (or row) in `Excel`.


2. `DataFrames`: those are spreadsheet-like tables that contain one or more `Series`.
    * It helps to think of `DataFrames` as tables (or spreadsheets) in `Excel`.



### `Series`


* `Series` is "similar" to a python `lists.`
  * An ordered collection of items of the **same datatype**.  
* `Series` contain an additional array of labels that are associated with each data entry.

| first_name | last_name | age | salary | owns_house | nb_cars |
|:---:|:---:|:---:|:---:|:---:|:---:|
| John | Smith | 35 | 162,321 | Yes | 2 |



### About the Data

- The examples below use cost of drugs prescribed to Medicare patients.

- The complete dataset is publicly available on the  Centers for Medicare & Medicaid Services ([`CMS` website](https://www.cms.gov/OpenPayments/Explore-the-Data/Dataset-Downloads.html)):

- This toy dataset contains the following columns:

| Column |Description|
|:----------|-----------|
| `unique_id`| A unique identifier for a Medicare claim to CMS |
| `doctor_id` | The Unique Identifier of the doctor who <br/> prescribed the medicine  |
| `specialty` | The specialty of the doctor prescribed the medicine |
| `medication` | The medication prescribed |
| `nb_beneficiaries` | The number of beneficiaries the <br/> medicine was prescribed to  |
| `spending` | The total cost of the medicine prescribed <br/>for the CMS |


<img src="https://www.dropbox.com/s/ha5s9t1n7yl0yx4/medicare_data.png?dl=1" alt="drawing" style="width:800px;"/>


### Example of Series as Rows and Columns

* It helps to think of Pandas `series` as either a row (or a column) of an Excel spreadsheet

<img src="https://www.dropbox.com/s/imt11t2xnw4f2d0/possible_series.png?dl=1" alt="drawing" style="width:500px;"/>




### Creating a `pandas` `Series`



```python
import pandas as pd
```

* To `create` a pandas `Series`, you can call the `Series` function and pass it a `list` of values and a `list` of labels (an `index`)

```python

>>> s =  pd.Series( data = [1234, 'DIAZEPAM', 3, '$32'],
                            index= ['id', 'med', 'ben', 'spen'])

>>> s
doctor_id               1234
medication          DIAZEPAM
nb_beneficiaries           3
spending                 $32
dtype: object
```


In [44]:
import pandas as pd
s =  pd.Series(data= [1234, 'DIAZEPAM', 3, '$32'], index= ['doctor_id', 'medication', 'nb_beneficiaries', 'spending'])
s

doctor_id               1234
medication          DIAZEPAM
nb_beneficiaries           3
spending                 $32
dtype: object

### Indexing `Series`

* You can access the data in `pandas` by index (position of the object) using the same approach seen in list indexing

```python
>>> s[1]
'DIAZEPAM'

>>> s[-1]
'$32'
```

* You can also access a value in the array using the data index

```python
>>> s["medication"]
'DIAZEPAM'

s["spending"]
'$32'
```
* Based on  the above, it is fair to think of a `Series` as a hybrid between lists an dictionaries

In [45]:
print(s[1])
print(s["medication"])

DIAZEPAM
DIAZEPAM


### Subsetting `Series`

- As with lists, subsetting `Series` can be carried out using the range operator ":"

```python
>>> s[0:3]
doctor_id               1234
medication          DIAZEPAM
nb_beneficiaries           3
dtype: object
```

### Subsetting `Series` - cont's

- Subsetting `Series` can also be done with lists of indices

```python
>>> s[[0,2,1]]
doctor_id               1234
nb_beneficiaries           3
medication          DIAZEPAM
dtype: object
```

- Note that the line above contains two sets of square brackets `[[ ]]`, the first is the indexing operator, the second (inner) is the delimiter for the list.
    

In [46]:
s[0:3]

doctor_id               1234
medication          DIAZEPAM
nb_beneficiaries           3
dtype: object

In [47]:
s[[0,2]]

doctor_id           1234
nb_beneficiaries       3
dtype: object

### Subsetting Series - Cont'd

* Subsetting series can also be done using lists of labels

```python

>>> s[["doctor_id", "nb_beneficiaries"]]
doctor_id           1234
nb_beneficiaries       3
dtype: object
```

* Note that line above also contains two sets of square brackets `[[ ]]`, the first is the indexing operator, the second, inner, set is the delimiter for the list.
    

In [48]:
s[["doctor_id", "nb_beneficiaries"]]

doctor_id           1234
nb_beneficiaries       3
dtype: object

### `DataFrames`

* It helps to think to `DataFrames` are spreadsheet-like tables made of sets of ordered Series

<img src="https://www.dropbox.com/s/cc6037l3lz0bc03/df_as_series.png?dl=1" alt="drawing" style="width:500px;"/>

* The example above illustrates a collection of column `Series` as a `DataFrame`, but the analogy holds for row `Series`

### Reading a `DataFrame` From a File

* A `DataFrame` can be created by reading data from a file

  * `pandas` supports many input formats, including `Excel`, `CSV`, `TSV`, `SAS`, `STATA`, etc.

* We can read the example tab-delimited (TSV) spending table using:
    
```python
>>> spending_df  = pd.read_csv("data/spending.csv")
```

* Jupyter beautifies `DataFrame`s by printing them  as `HTML` tables.

  * Using `print` prints them as text instead. So try to print them by including the name as the last statement of the cell.
  

In [2]:
import pandas as pd 
spending_df  = pd.read_csv("https://www.dropbox.com/s/cgxt1qt5xzx9a3x/spending.csv?dl=1")

spending_df

unique_id   doctor_id        specialty     medication  nb_beneficiaries  \
0  AB789982  1952310666       Psychiatry     CLONAZEPAM               226   
1  AV967778  1952310666       Psychiatry       DIAZEPAM               103   
2  CC128705  1298765423       Cardiology        NADOLOL                13   
3  GH890091  1346358827           Family    HYDROCODONE               331   
4  YY219322  1548247315       Psychiatry     ALPRAZOLAM                28   
5  YY190561  1548247315       Psychiatry     GABAPENTIN                86   
6  YY572610  1548247315       Psychiatry    MIRTAZAPINE               191   
7  PL346720  1326175365           Family  OXYCODONE HCL                87   
8  GZ129032  1518970284  Hemato-oncology        DIGOXIN                54   

     spending  
0   $1,848.88  
1     $662.87  
2    $1968.27  
3   $8,511.14  
4   $1,964.49  
5   $1,807.16  
6   $3,131.96  
7  $12,881.04  
8   $3,766.34

### Dimensions of the DataFrame 

* Some useful properties and methods of of the `DataFrame`

  * `shape`: describes the number of lines and the number of columns in your `DataFrame`.
  * `head(n=5)`: prints the first `n` lines of your `DataFrame`
  * `tail(n=5)`: prints the last `n` lines of your `DataFrame`

```

In [5]:
spending_df.head(2)

unique_id   doctor_id   specialty  medication  nb_beneficiaries   spending
0  AB789982  1952310666  Psychiatry  CLONAZEPAM               226  $1,848.88
1  AV967778  1952310666  Psychiatry    DIAZEPAM               103    $662.87

In [6]:
spending_df.tail(2)

unique_id   doctor_id        specialty     medication  nb_beneficiaries  \
7  PL346720  1326175365           Family  OXYCODONE HCL                87   
8  GZ129032  1518970284  Hemato-oncology        DIGOXIN                54   

     spending  
7  $12,881.04  
8   $3,766.34

In [51]:
spending_df.shape

(9, 6)

### `DataFrame`, Indexes and Columns

* As opposed to `Series` which only have indexes, `DataFrame`s have `Indexes` (labels and positions) and `Column` (labels and positions).

  * In the example above, since a row label was not explicitly provided, the labels for the rows are the same as the indices, i.e., 0 through 8.


* They can, therefore, be indexed by row index, row label, column index, column label, or by a combination of both


<img src="https://www.dropbox.com/s/fi9elbiyaealjt3/df_index_cols.png?dl=1" alt="drawing" style="width:500px;"/>


In [9]:
spending_df.index

RangeIndex(start=0, stop=9, step=1)

### Index Location

- Row indexing can be carried out by passing the `iloc` (index location) operator a single index or a list of indexes.

  -  The `iloc` operator uses `[]` instead of the `()` that methods and functions use.


```python    
>>> spending_df.iloc[3]
```

or

```python    
>>> spending_df.iloc[[1,5]]
```

- When a single index is given, `pandas` returns a `Series`
  - Remember that a row is merely a `Series`
  
- When a list of indices is given, `pandas` returns another `DataFrame` 
  - The returned `DataFrame` is a subset of the original
  

In [52]:
# Returns row, or Series
spending_df.iloc[3]

unique_id              GH890091
doctor_id            1346358827
specialty                Family
medication          HYDROCODONE
nb_beneficiaries            331
spending              $8,511.14
Name: 3, dtype: object

In [53]:
# Returns two rows as a DataFrame

spending_df.iloc[[1,5]]

unique_id   doctor_id   specialty  medication  nb_beneficiaries   spending
1  AV967778  1952310666  Psychiatry    DIAZEPAM               103    $662.87
5  YY190561  1548247315  Psychiatry  GABAPENTIN                86  $1,807.16

### Indexing by Rows and Column indexes 

* You can pass `iloc` a combination of row and column indexes using the following construct

```python
 data_frame.iloc[ row_index_info, column_index_info]
```
* `row_index_info` and `column_index_info` can be either a single index or a list of indexes.


Example:

```python
>>>  spending_df.iloc[3, 3]
'HYDROCODONE'

>> spending_df.iloc[2, [1,3]]
doctor_id     1298765423
medication       NADOLOL
Name: 2, dtype: object

>>>  spending_df.iloc[[2,4], [1,3]]
```
|  |      medication |   spending |
|:--|:-----------------|:------------|
|2 |   HYDROCODONE   | \$8,511.14  |
|4 | OXYCODONE HCL   | \$12,881.04 |


### Reading Tables with Index Labels

*  Rather than using the default integer index label created by `pandas`, a `DataFrame` can be indexed using one or more columns of data.

*  The index need not be an integer and can consist of any type.

![](https://www.dropbox.com/s/qb62v2iov9frp38/default_index.png?dl=1)

In [10]:
spending_df.head()


unique_id   doctor_id   specialty   medication  nb_beneficiaries   spending
0  AB789982  1952310666  Psychiatry   CLONAZEPAM               226  $1,848.88
1  AV967778  1952310666  Psychiatry     DIAZEPAM               103    $662.87
2  CC128705  1298765423  Cardiology      NADOLOL                13   $1968.27
3  GH890091  1346358827      Family  HYDROCODONE               331  $8,511.14
4  YY219322  1548247315  Psychiatry   ALPRAZOLAM                28  $1,964.49

### Specifying and Index Labels

* We can use any column of the data to label the indexes by passing the label of the column you want to use to `pandas` `read_csv()` function.
  - For example, we can use the `unique_id` column as index labels 

```python
spending_df  = pd.read_csv( "data/spending.csv", 
                                  index_col=["unique_id"] )
```

<img src="https://www.dropbox.com/s/wmixxjn6dol14b6/custom_index.png?dl=1" alt="drawing" style="width:600px;"/>



In [11]:

spending_df  = pd.read_csv( "https://www.dropbox.com/s/cgxt1qt5xzx9a3x/spending.csv?dl=1", 
                                  index_col=["unique_id"] )

# after specifying the index column
# note that index is now unique_id

spending_df.head()

doctor_id   specialty   medication  nb_beneficiaries   spending
unique_id                                                                  
AB789982   1952310666  Psychiatry   CLONAZEPAM               226  $1,848.88
AV967778   1952310666  Psychiatry     DIAZEPAM               103    $662.87
CC128705   1298765423  Cardiology      NADOLOL                13   $1968.27
GH890091   1346358827      Family  HYDROCODONE               331  $8,511.14
YY219322   1548247315  Psychiatry   ALPRAZOLAM                28  $1,964.49

### Indexing DataFrames Columns using Labels

* Accessing data in the `DataFrame` can also be done using indexes or column label.


* Indexing a column takes a column's label or a list of index labels

```python
spending_df["nb_beneficiaries"]
```
or 
```python
spending_df[["doctor_id", "nb_beneficiaries"]]
```



In [55]:
spending_df[["doctor_id", "nb_beneficiaries"]]

doctor_id  nb_beneficiaries
unique_id                              
AB789982   1952310666               226
AV967778   1952310666               103
CC128705   1298765423                13
GH890091   1346358827               331
YY219322   1548247315                28
YY190561   1548247315                86
YY572610   1548247315               191
PL346720   1326175365                87
GZ129032   1518970284                54

### Indexing DataFrames Columns using Labels - Cont'd
* When, a single label is given, `pandas` returns a `Series`
  * Remember that a column is simply a `Series`
  
  
* When, a list of labels is given, `pandas` returns another `DataFrame` 

  * The returned `DataFrame` is a subset of the original

In [56]:
spending_df["nb_beneficiaries"]

unique_id
AB789982    226
AV967778    103
CC128705     13
GH890091    331
YY219322     28
YY190561     86
YY572610    191
PL346720     87
GZ129032     54
Name: nb_beneficiaries, dtype: int64

In [57]:
spending_df[["doctor_id", "nb_beneficiaries"]]

doctor_id  nb_beneficiaries
unique_id                              
AB789982   1952310666               226
AV967778   1952310666               103
CC128705   1298765423                13
GH890091   1346358827               331
YY219322   1548247315                28
YY190561   1548247315                86
YY572610   1548247315               191
PL346720   1326175365                87
GZ129032   1518970284                54

### Indexing `DataFrames` Rows Using Labels

* Use `loc` (location) operator to index a row using a label


  -  The `loc` operator uses `[]` instead of the `()` that methods and functions use.
 
*  `loc` can take a single label or a list of labels 

```python    
spending_df.loc["AV967778"]
```

or

```python    
spending_df.loc[["AV967778","YY219322"]]
```


In [14]:
spending_df.loc[["AV967778", "YY219322"], ["nb_beneficiaries", "specialty"]]

nb_beneficiaries   specialty
unique_id                              
AV967778                103  Psychiatry
YY219322                 28  Psychiatry

### Indexing `DataFrames` Rows Using Labels - cont'd

* When a single index is given, `pandas` returns a `Series`.
  - Remember that a row is simply a `Series`.
  
* When a list of indices is given, `pandas` returns another `DataFrame`
  - The returned `DataFrame` is a subset of the original

In [58]:
spending_df.loc["AV967778"]

doctor_id           1952310666
specialty           Psychiatry
medication            DIAZEPAM
nb_beneficiaries           103
spending               $662.87
Name: AV967778, dtype: object

In [59]:
spending_df.loc[["AV967778","YY219322"]]

doctor_id   specialty  medication  nb_beneficiaries   spending
unique_id                                                                 
AV967778   1952310666  Psychiatry    DIAZEPAM               103    $662.87
YY219322   1548247315  Psychiatry  ALPRAZOLAM                28  $1,964.49


##### Indexing review:

| Syntax                |   Meaning    |
|:--------------------------|:-------------------|
| `dataframe["col_name"]`    |  Return `Series` of "col_name"'s  |
| `dataframe[["col_name_1", "col_name_2"] ]`    |  Returns `DataFrame` with  "col_name_1" and "col_name_2"|
| `dataframe.loc["label"]`| returns entry indexed by "some_label" |
| `dataframe.loc["label", ["col_3", 'col_5']]`| return entry indexed by "some_label", subsets columns to only "col_3" and "col_5" |
| `dataframe.loc[["label_1", "label_2"], ['col_3', 'col_5']]`| returns lines with indices "label_1" and "label", subsets columns to only "col_3" and "col_5" |
| `dataframe.iloc[23, [0, 1]]`| returns line 23, and only values of columns 0 and 1 |
| `dataframe.iloc[[1,2], [0, 1]]`| returns lines 1 and 2, and only values of columns 0 and 1 |


In [17]:
spending_df["specialty"]

unique_id
AB789982         Psychiatry
AV967778         Psychiatry
CC128705         Cardiology
GH890091             Family
YY219322         Psychiatry
YY190561         Psychiatry
YY572610         Psychiatry
PL346720             Family
GZ129032    Hemato-oncology
Name: specialty, dtype: object

In [18]:
# We can, and we do often, index using Booleans
spending_df["specialty"] == "Psychiatry"

unique_id
AB789982     True
AV967778     True
CC128705    False
GH890091    False
YY219322     True
YY190561     True
YY572610     True
PL346720    False
GZ129032    False
Name: specialty, dtype: bool

In [22]:
is_psychiatry = (spending_df["specialty"] == "Psychiatry") & (spending_df["nb_beneficiaries"] > 100)

In [23]:
is_psychiatry

unique_id
AB789982     True
AV967778     True
CC128705    False
GH890091    False
YY219322    False
YY190561    False
YY572610     True
PL346720    False
GZ129032    False
dtype: bool

In [24]:
spending_df[is_psychiatry]

doctor_id   specialty   medication  nb_beneficiaries   spending
unique_id                                                                  
AB789982   1952310666  Psychiatry   CLONAZEPAM               226  $1,848.88
AV967778   1952310666  Psychiatry     DIAZEPAM               103    $662.87
YY572610   1548247315  Psychiatry  MIRTAZAPINE               191  $3,131.96

### Pandas cheat sheet
* see full cheat shee on the Miro Board

![](https://www.dropbox.com/s/bi7fqd9sz6bc2f7/pandas_summarize.png?dl=1)

In [91]:
spending_df["nb_beneficiaries"].sum()


1119

In [92]:
def dollar_to_flaot(x):
    """
     convert a `$` figure to a float
     ex. the string $1,178.29 is converted into the float 1278.29
    """
    x = x.replace(",", "")
    x = x.replace("$", "")
    
    return float(x)

In [93]:
spending_df["spending"].apply(dollar_to_flaot)


unique_id
AB789982     1848.88
AV967778      662.87
CC128705     1968.27
GH890091     8511.14
YY219322     1964.49
YY190561     1807.16
YY572610     3131.96
PL346720    12881.04
GZ129032     3766.34
Name: spending, dtype: float64

In [94]:
spending_df["sending_cleaned"] = spending_df["spending"].apply(dollar_to_flaot)
spending_df.head()

doctor_id   specialty   medication  nb_beneficiaries   spending  \
unique_id                                                                     
AB789982   1952310666  Psychiatry   CLONAZEPAM               226  $1,848.88   
AV967778   1952310666  Psychiatry     DIAZEPAM               103    $662.87   
CC128705   1298765423  Cardiology      NADOLOL                13   $1968.27   
GH890091   1346358827      Family  HYDROCODONE               331  $8,511.14   
YY219322   1548247315  Psychiatry   ALPRAZOLAM                28  $1,964.49   

           sending_cleaned  individual_cost  individual_cost_cdn  
unique_id                                                         
AB789982           1848.88        10.307915            12.987973  
AV967778            662.87         8.108895            10.217208  
CC128705           1968.27       190.770785           240.371189  
GH890091           8511.14        32.398902            40.822616  
YY219322           1964.49        88.402050           111.386583

In [95]:
spending_df["individual_cost"] =  spending_df["sending_cleaned"] / spending_df["nb_beneficiaries"]
spending_df.head()

doctor_id   specialty   medication  nb_beneficiaries   spending  \
unique_id                                                                     
AB789982   1952310666  Psychiatry   CLONAZEPAM               226  $1,848.88   
AV967778   1952310666  Psychiatry     DIAZEPAM               103    $662.87   
CC128705   1298765423  Cardiology      NADOLOL                13   $1968.27   
GH890091   1346358827      Family  HYDROCODONE               331  $8,511.14   
YY219322   1548247315  Psychiatry   ALPRAZOLAM                28  $1,964.49   

           sending_cleaned  individual_cost  individual_cost_cdn  
unique_id                                                         
AB789982           1848.88         8.180885            12.987973  
AV967778            662.87         6.435631            10.217208  
CC128705           1968.27       151.405385           240.371189  
GH890091           8511.14        25.713414            40.822616  
YY219322           1964.49        70.160357           111.386583

In [96]:
def us_to_cdn(x):
    return x * 1.26


In [97]:
spending_df["individual_cost_cdn"] = spending_df["individual_cost"].apply(us_to_cdn)
spending_df

doctor_id        specialty     medication  nb_beneficiaries  \
unique_id                                                                 
AB789982   1952310666       Psychiatry     CLONAZEPAM               226   
AV967778   1952310666       Psychiatry       DIAZEPAM               103   
CC128705   1298765423       Cardiology        NADOLOL                13   
GH890091   1346358827           Family    HYDROCODONE               331   
YY219322   1548247315       Psychiatry     ALPRAZOLAM                28   
YY190561   1548247315       Psychiatry     GABAPENTIN                86   
YY572610   1548247315       Psychiatry    MIRTAZAPINE               191   
PL346720   1326175365           Family  OXYCODONE HCL                87   
GZ129032   1518970284  Hemato-oncology        DIGOXIN                54   

             spending  sending_cleaned  individual_cost  individual_cost_cdn  
unique_id                                                                     
AB789982    $1,848.88          1848.88         8.180885            10.307915  
AV967778      $662.87           662.87         6.435631             8.108895  
CC128705     $1968.27          1968.27       151.405385           190.770785  
GH890091    $8,511.14          8511.14        25.713414            32.398902  
YY219322    $1,964.49          1964.49        70.160357            88.402050  
YY190561    $1,807.16          1807.16        21.013488            26.476995  
YY572610    $3,131.96          3131.96        16.397696            20.661097  
PL346720   $12,881.04         12881.04       148.057931           186.552993  
GZ129032    $3,766.34          3766.34        69.747037            87.881267